enhancement/14010-new-demo-technical-indicators

karolkolodziej opened the pull request:
Feature #14010, added the demo with all technical indicators.

Edited demo: http://jsfiddle.net/BlackLabel/czpj5h74/

circleci-mu commented:
## Visual test results - No difference found

---
<details>
<summary>Samples changed</summary><p>

| Change type | Sample |
| --- | --- |
|  Added | samples/stock/demo/all-indicators/demo.js |

</p>
</details>

circleci-mu commented:
### File size comparison
No differences found

raf18seb commented on samples/stock/demo/all-indicators/demo.html:
```diff
@@ -0,0 +1,65 @@
+<script src="https://code.highcharts.com/stock/modules/exporting.js"></script>
+<script src="https://code.highcharts.com/stock/indicators/indicators-all.js"></script>
+
+<div class='main'>
+  <div class='selectors'>
+    <div class="left">
+      <label class='left__labels' for="overlays">Overlays:</label>
+      <select class='left__select' id="overlays">
+        <option value="none">None</option>
```
I haven't seen many HTML tags in our official demos. We could consider using Highcharts tools e.g. `H.createElement` and then we could position 'selects' using Highcharts coordinates (without CSS). @pawelfus what do you think?

pawelfus replied:
I think simple HTML in the demo is fine. For example Periodic table: http://utils.highcharts.local/samples/#view/highcharts/series-heatmap/periodic-table

raf18seb commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,193 @@
+            name: 'Volume',
+            data: volume,
+            yAxis: 1
+        }]
+    });
+
+    document.getElementById("overlays").addEventListener("change", function (e) {
```
I would put this code either into the chart's callback or load event.

raf18seb commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,193 @@
+        if (chart.series[2] && chart.series[2].yAxis.options.index === 0) {
+            chart.series[2].destroy(false);
+        }
+        if (chart.series[3] && chart.series[3].yAxis.options.index === 0) {
+            chart.series[3].destroy(false);
+        }
```
What is the purpose of this if statement?

karolkolodziej replied:
User can add first an oscillator (which will be series with the index 2), then can add an overlay (which will be with index 3). The only thing (as far as I can see) which you can use to distinguish them is the yAxis index.
The main purpose for that statements is to detect which series is assigned to which axis and destroy the correct one.
Is there any room for improvement?

raf18seb commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,193 @@
+                height: '25%'
+            });
+        } else {
+            if (seriesToAdd === 'aroon') {
+                chart.addSeries({
+                    yAxis: 3,
+                    type: 'aroon',
+                    linkedTo: 'aapl',
+                    color: 'green',
+                    lineWidth: 1,
+                    showInLegend: true,
+                    aroonDown: {
+                        styles: {
+                            lineColor: 'red'
+                        }
+                    }
+                }, false);
+            }
+
+            if (seriesToAdd !== 'aroon') {
+                chart.addSeries({
+                    type: seriesToAdd,
+                    linkedTo: 'aapl',
+                    yAxis: 3,
+                    showInLegend: true
+                }, false);
+            }
```
We could put it into one `if-else` statement.

raf18seb commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,193 @@
+        volume = [],
+        dataLength = data.length,
+
+        i = 0;
+
+    for (i; i < dataLength; i += 1) {
```
We can shorten this to one line.

pawelfus commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,193 @@
+        title: {
+            text: 'AAPL Historical'
+        },
+        subTitle: {
```
`subtitle`

pawelfus commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,193 @@
+        if (chart.yAxis.length === 3) {
+            chart.yAxis[0].update({
+                height: '55%'
+            });
```
When using multiple updates, set `redraw=false` (the same applies to the `update()` and `addAxis()` below)

pawelfus commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,193 @@
+                        styles: {
+                            fill: 'rgba(0, 0, 255, 0.1)'
+                        }
+                    }
```
Do we need to define all styles for lines in IKH?

karolkolodziej replied:
No, we don't have to, actually most of them (especially the fill) are set by default.

pawelfus commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,193 @@
+            selected: 2
+        },
+        tooltip: {
+            split: true
```
This is the default, isn't it?

pawelfus commented on samples/stock/demo/all-indicators/demo.css:
```diff
@@ -0,0 +1,42 @@
+.selectors {
+	height: 20px;
+	font-size: 20px;
+	display: flex;
```
+ `flex` is tricky when comes to auto-resizing the chart

karolkolodziej replied:
I removed the `flex` property and I used some different style attributes.

raf18seb reviewed (approved):
Looks good, thanks!

pawelfus commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,147 @@
+                        color: 'green',
+                        lineWidth: 1,
+                        showInLegend: true,
+                        aroonDown: {
```
Can we simplify it? I mean, let's treat `aroon` as any other tech. indicator, without defining styles (just use defaults)

pawelfus commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,147 @@
+            chart.redraw();
+        });
+
+        document.getElementById("oscilators").addEventListener("change", function (e) {
```
Sorry for missing this earlier, just a typo: "oscillators"

pawelfus reviewed (commented):
Can we change the demo, I mean the way it works? For example:

1. Let's init chart and select by default one overlay and one oscillator (with its own yAxis)
2. Remove `None` option from dropdowns. That simplifies a lot demo (managing yAxis heights, indexes, etc)
3. Add `id` to overlay and oscillator series. So when changing indicator, we simply use `let series = chart.get('series-id'); if (series) { series.remove(false) }` then add new series

karolkolodziej commented:
I've changed that logic and simplified the demo as you mentioned.

The demo looked strange in utils because of the CSS, I also fixed that.
Updated demo: http://jsfiddle.net/BlackLabel/0qnjfxtv/

sebastianbochan reviewed (changes requested):
- change class names
- replace px with em (font-size)
- remove relative / top and use margins / display-inline etc.

raf18seb commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    // create the chart
+    const chart = Highcharts.stockChart('container', {
```
`const` will not work in the old browsers. We can remove it (see the comment below).

raf18seb commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,107 @@
+            showInLegend: true,
+            yAxis: 2
+        }]
+    }, function () {
```
Instead of declaring a `const chart` variable, we can use `chart` as function's argument like
```js
}, function (chart) {
```

raf18seb commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,107 @@
+            yAxis: 2
+        }]
+    }, function () {
+        document.getElementById("overlays").addEventListener("change", function (e) {
+            const series = chart.get('overlay');
```
Another `const`. Please, find all of them. The same with double quotes.

raf18seb replied:
Mixing single and double quotes. Prefer single `"` -> `'`

raf18seb commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,107 @@
+                    type: e.target.value,
+                    linkedTo: 'aapl',
+                    id: 'overlay',
+                    yAxis: 0,
```
Is it necessary? `yAxis: 0` is by default. Or maybe this indicator doesn't stick to the first axis by default?

raf18seb commented on samples/stock/demo/all-indicators/demo.js:
```diff
@@ -0,0 +1,107 @@
+                    linkedTo: 'aapl',
+                    id: 'overlay',
+                    yAxis: 0,
+                    showInLegend: true
```
There are many `showInLegend: true`. If all series should be shown in the legend, you could define it in `plotOptions.series`.

raf18seb commented on samples/stock/demo/all-indicators/demo.details:
```diff
@@ -0,0 +1,10 @@
+---
+name: All technical indicators
+authors:
+  - Karol Kołodziej
```
By the occasion - don't you want to replace `ł` with `l` as we discussed?

raf18seb commented on samples/stock/demo/all-indicators/demo.html:
```diff
@@ -0,0 +1,64 @@
+<script src="https://code.highcharts.com/stock/indicators/indicators-all.js"></script>
+
+<div class="main-wrapper">
+  <div class="selectors-container">
```
You can leave it as it is, but do we need this `<div class="selectors-container">`? I don't see any good reasons (maybe there is one) why this
```html
<div class="a">
    <div class="b">
        <p></p>
        <p></p>
        <p></p>
    </div>
</div>
```
cannot be replaced with (with corrected CSS styles and classes maybe)
```html
<div class="a b">
    <p></p>
    <p></p>
    <p></p>
</div>
```

Just a suggestion for the future.

karolkolodziej replied:
`main-wrapper` class is to... well.. wrap the whole selectors and the chart together as it might be a part of some bigger design/site

raf18seb commented on samples/stock/demo/all-indicators/demo.html:
```diff
@@ -0,0 +1,64 @@
+  <div class="selectors-container">
+    <div class="lhs-col">
+      <label for="overlays">Overlays:</label>
+      <select class='left-select' id="overlays">
```
In this case, a double quote is preferable `"`.

raf18seb reviewed (approved):
Thanks!

pawelfus reviewed (approved):
Thanks!

pawelfus commented on samples/stock/demo/all-indicators/demo.html:
```diff
@@ -0,0 +1,64 @@
+    <div class="rhs-col">
+      <label for="oscillators">Oscillators:</label>
+      <select class="right-select" id="oscillators">
+        <option value="none">None</option>
```
Accepted, and then realized: There is no `none` technical indicator :)

pawelfus commented on samples/stock/demo/all-indicators/demo.html:
```diff
@@ -0,0 +1,64 @@
+        <option value="apo">Absolute price indicator</option>
+        <option value="ad">A/D (Accumulation/Distribution)</option>
+        <option value="aroon">Aroon</option>
+        <option value="aroonoscillator">Aroon oscilator</option>
```
Oscillator :)

pawelfus commented on samples/stock/demo/all-indicators/demo.html:
```diff
@@ -0,0 +1,64 @@
+        <option value="aroon">Aroon</option>
+        <option value="aroonoscillator">Aroon oscilator</option>
+        <option value="atr">ATR (Average True Range)</option>
+        <option value="ao">Awesome oscilator</option>
```
Oscillator

pawelfus reviewed (changes requested):
minor polishing

pawelfus reviewed (approved):
Thanks!

pawellysy reviewed (approved):
Looks good!

TorsteinHonsi reviewed (approved):
Thanks, nice demo! I did some final CSS polishing, now merging.
